infer/resource_test.go

```diff
@@ -449,3 +449,20 @@ func TestHydrateFromState(t *testing.T) {
 		}),
 	))
 }
+
+type inputForDefaultCheck struct {
+	P1 string `pulumi:"str,optional"`
+}
+
+func (i *inputForDefaultCheck) Annotate(a Annotator) {
+	a.SetDefault(&i.P1, "default")
+}
+
+func TestDefaultCheckAppliesDefaultValues(t *testing.T) {
```

Comment:
This is a good unit test. It would be great if we could parameterize by if `Check` if implemented (and just returns `DefaultCheck`) vs left as default. We could then assert that they behave identically.

We can leave that to a future exercise.

Reply:
I implemented this in the latest commit. Sharing `DefaultCheck` required separating decoding from applying defaults, and didn't really resulted in any LOC savings, but at least the logic is now guaranteed to be the same.